### ISSUE TITLE
Instant precision should be retained in ObjectMapper#readTree (add `JsonNodeFeature.USE_BIG_DECIMAL_FOR_FLOATS` to force use of `BigDecimal` for `JsonNode` reads)

### ISSUE THREAD

**Issue by jzheaux** [FasterXML/jackson-modules-java8#307]
**Describe the bug**

`ObjectMapper#readValue(String)` seems to preserve all nine digits of an `Instant`'s decimal, but `ObjectMapper#readTree(JsonParser)` does not.

**Version information**

Jackson 2.12.1

**To Reproduce**

```java
ObjectMapper mapper = new ObjectMapper();
{
	mapper.registerModule(new JavaTimeModule());
}

Instant issuedAt = Instant.ofEpochSecond(1234567890).plusNanos(123456789);
Map<String, Instant> attributes = new HashMap<>();
{
	attributes.put("iat", issuedAt);
}

@Test
void passes() throws Exception {
	String serialized = mapper.writeValueAsString(attributes);
	Map<String, Instant> deserialized = mapper.readValue(serialized, new TypeReference<>() {});
	assertThat(deserialized.get("iat")).isEqualTo(issuedAt);
}

@Test
void fails() throws Exception {
	String serialized = mapper.writeValueAsString(attributes);
	JsonParser parser = mapper.createParser(serialized);
	JsonNode mapNode = mapper.readTree(parser);
	Map<String, Object> deserialized = new LinkedHashMap<>();
	Iterable<Map.Entry<String, JsonNode>> fields = mapNode::fields;
	for (Map.Entry<String, JsonNode> field : fields) {
		deserialized.put(field.getKey(), mapper.readValue(field.getValue().traverse(mapper), Instant.class));
	}
	assertThat(deserialized.get("iat")).isEqualTo(issuedAt);
}
```

The second test fails with:

```text
Expecting:
 <2009-02-13T23:31:30.123456700Z>
to be equal to:
 <2009-02-13T23:31:30.123456789Z>
but was not.
Expected :2009-02-13T23:31:30.123456789Z
Actual   :2009-02-13T23:31:30.123456700Z
```

**Additional context**

The failure case is derived from [a custom deserializer](https://github.com/spring-projects/spring-security/blob/master/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/jackson2/UnmodifiableMapDeserializer.java) in our project.

Ideally, I'd like `readTree` to retain the `Instant`'s precision in the same way that `ObjectMapper#readValue` does when it deserializes a `Map` that contains `Instant`s, though I'd be happy with advice on how to change my approach in my deserializer instead.

I realize also that `USE_BIG_DECIMAL_FOR_FLOATS` addresses this issue, but I'd prefer to avoid that, since others import [our module](https://github.com/spring-projects/spring-security/blob/master/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/jackson2/OAuth2ClientJackson2Module.java).

**Comment by Logic-32** [comment]
Sorry.  Perhaps "default" was the wrong word.  What I mean is:

- On the happy-path that works as expected, the code goes through: https://github.com/FasterXML/jackson-core/blob/8b87cc1a96f649a7e7872c5baa8cf97909cabf6b/src/main/java/com/fasterxml/jackson/core/base/ParserBase.java#L863
    - This path explicitly uses `BigDecimal` throughout the whole process and allows for the retention of nanosecond precision.
 - On the sad-path, using `JsonNode`, you go through: https://github.com/FasterXML/jackson-databind/blob/db623a8a85834f59f5b6331435b80150050ef841/src/main/java/com/fasterxml/jackson/databind/node/TreeTraversingParser.java#L290
    - This path allegedly uses a `BigDecimal` too.  However, at this point, the default behavior is to have a `DoubleNode` which loses the nanosecond precision.

So, what I mean is that `ParserBase` has the correct behavior during parsing without changing any settings.  But, the code that parses a `JsonNode` requries the extra setting to operate as expected.

With that said, I kinda understand why you moved this case to `jackson-modules-java8` however, I'm curious what can be done in/around the `InstantDeserializer` to allow `JsonNode` to retrain the correct precision without the `USE_BIG_DECIMAL_FOR_FLOATS` feature?

**Comment by cowtowncoder** [comment]
Unfortunately I don't see another solution, aside possibly adding separate `JsonNodeFeature` to allow forcing of `BigDecimal` just for `JsonNode` and not generally.

The issue is that although in theory we could change handling of `JsonNode` to use "deferred" number value (accessible via `JsonParser.getNumberValueDeferred()`), `JsonNode` subtypes are strongly typed with eager values -- and changing that could have major compatibility consequences.
This means that when constructing a `JsonNode` based tree, physical number type will be fixed, and there is no way recover (by re-parsing) from original textual number representation into more precise type, if `DoubleNode` has been created.

So the question is whether processing itself absolutely requires `JsonNode` as intermediate type. Then again, binding into `Object` faces similar problem: the only structure that retains deferred type is `TokenBuffer` used internally for buffering. It is available externally, but is not very convenient for any actual processing, unlike `JsonNode`.

**Comment by Logic-32** [comment]
> Unfortunately I don't see another solution, aside possibly adding separate JsonNodeFeature to allow forcing of BigDecimal just for JsonNode and not generally.

I'm not immediately opposed to this idea.  In general, I'd be surprised if anyone expected `JsonNode` to exhibit different parsing results from what `ParserBase` offers.  Meaning, if this is a feature, I'd like to see it enabled by default so the decision about precision can be made as late as possible.

> So the question is whether processing itself absolutely requires JsonNode as intermediate type.

In our case, we absolutely do not need `JsonNode`.  In fact, I lobby against using `JsonNode` in almost all cases.  It has a place but people lean on it too hard imo.  With that said, I also can't justify the cost to refactor our current, legacy code to not use it :(

**Comment by cowtowncoder** [comment]
Yeah the idea of adding the new feature would specifically be to allow forcing `BigDecimal` only for `JsonNode`, not other processing (for `Number` or `Object`, specifically) -- that is, can leave `DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS)` disabled.

**Comment by Logic-32** [comment]
I'm not sure if that'll leave some other corner cases around but am not immediately opposed to it :)  Afterall, my arguments against JsonNode in general is that it is not an efficient approach to deserialization.  May as well at least make it "accurate" ;)

**Comment by cowtowncoder** [comment]
`JsonNode` is interesting: for some use case it is pretty performant (when it is the main abstraction used, and not used as just source for converting to something else, i.e. unnecessary transformations). But it is also easy to overuse.

Be that as it may, the intent with possible addition would be to give more granular configurability -- I do not think there is a way to automatically prevent the problem as per my earlier explanation & backwards compatibility constraints.

**Comment by cowtowncoder** [comment]
No real update, other than mention the possibility of adding a new `JsonNodeFeature`.

Also, adding a (failing) test based on code here, PR for, would be something useful.

**Comment by cowtowncoder** [comment]
Since no one else did it, I added test (in bit simplified form), in case someone has time to dig into this in future (but as noted, can't be fixed by this module alone).

**Comment by cowtowncoder** [comment]
Fixed by introduction of `JsonNodeFeature.USE_BIG_DECIMAL_FOR_FLOATS` (in 2.19.0); closing.
